### ISSUE TITLE
Graph improvement + sparse discussion

### ISSUE THREAD

**Issue by firasdib** [eclipse/deeplearning4j#3997]
Hello,

I've been working on a project for the last couple of months which has required a lot of flexibility from the library. Today I have finally hit a snag which I'm not entirely sure how to solve.

The issue is two-fold, and I will discuss both below.

## Graph
I've been using the graph to construct my network and its been working well so far. However, I now need to map multiple inputs to a single layer, receive multiple outputs, and after that pass them to a vertex (to merge, or what have you).

As of this writing this is not possible, and one potential (untested) solution is that I simply duplicate the layer that is going to receive multiple inputs. I have illustrated my problem with a few images below.

### Ideal model
Here is an illustration of how I would ideally like to model it. 
In code this would be something like:

```java
graph.addLayer("myLayerName", ..., "input1", "input2", "input3", "input4");
```

![ideal_model](https://user-images.githubusercontent.com/1335165/30003249-b7d87660-90ba-11e7-84f4-0b8c1fff12f9.png)

Here is a python counter example where something similar has been done: https://github.com/xmb-cipher/fofe-ner/blob/master/fofe_mention_net.py#L1823-L1964

### Potential workaround
Here is an illustration of a potential workaround. I have not been able to test it yet (due to issue number two).

In code this would look something like:

```java
Layer layer = ...;

graph.addLayer("layer1", layer, "input1");
graph.addLayer("layer2", layer, "input2");
graph.addLayer("layer3", layer, "input3");
graph.addLayer("layer4", layer, "input4");
```

Note that I'm using the same reference for the layer. This is depicted as a dotted line in the image below.

![workaround](https://user-images.githubusercontent.com/1335165/30003274-071ad6e6-90bb-11e7-9272-9b40bc294e83.png)

## Sparse
The second issue is sparse support. These layers that you see in the images above are supposed to be projection layers. They will be the first layer in the graph which will take a sparse input and convert it into a dense one (simple multiplication with layer weights). The weights are initialized with a pre-trained word2vec model.

To clarify, the input needs to be sparse, the output will be fixed-length and dense.

The input looks something like this (but much, much larger):
```
[0 0 0 0.25 0 0 0 0 0 0.7 0 0 1 0 0 0 0 0 0.2]
```

The reason this needs to be sparse is because these vectors would otherwise be huge, around 100-200k elements of which only a handful would be non-zero, and not fit in memory.

From the code I have managed to read and understand, it would seem that the layers expect a dense matrix as their input with no exception. One potential workaround could be to pad the vectors so they are all the same length as the longest one. Problem here is that there is no guarantee there won't be a huge vector and in that case we're still back with a huge matrix.

My end goal here was to create a custom ProjectionLayer, which works very similar to EmbeddingLayer, but graph multiple rows from the weight matrix (one for each non-zero entry) and multiplies it with the sparse vector.

Any and all feedback would be much appreciated.

ping @raver119 @agibsonccc @AlexDBlack

**Comment by AlexDBlack** [comment]
Many (but not all) multiple-input use cases can be handled via merging (and if necessary, subsetting in the layer).
Multiple outputs isn't currentyl supported - but in many cases, can be handled (worked around) by merging in the layer, and doing a subset operation.
Of course, this doesn't work if the output shapes don't match or can't be merged.
See also https://github.com/deeplearning4j/deeplearning4j/issues/3998

In your "potential workaround" section looks like you want parameter sharing? i.e., multiple identical layers, with one parameter vector.
This isn't currently supported. A workaround is to use StackVertex + UnstackVertex, which is functionally the same thing, and we've used this successfully for things like triplet embedding models.



As for sparse - yes, dense only currently - the restriction currently is the lack of sparse support in ND4J.
With regard to the idea of a custom ProjectionLayer, that could work: we could use what we might term a "pseudo-sparse" representation (N values would have 2N entries, (index, value) pairs, with say index=-1 used for not present). But to implement that efficiently, it would likely be necessary to push that down to the c++ level, for things like matrix multiplications.

**Comment by agibsonccc** [comment]
A few counter points here:
We have the sitting PR that could be merged with sparse with caveats depending on the needs for access here.

For the parameter sharing..if it ends up being a sequential operation in practice anyways, what consequences would there be if ops are sequential in the resolution anyways?

As for multi input/output based on internal discussions, I feel like we could do this in the near term to at least unblock certain kinds of operations.

**Comment by firasdib** [comment]
@AlexDBlack Regarding parameter sharing: would my proposed solution not work? The reference to the layer is identical, so I was hoping the graph would think its N different layers when in fact its the same, just with different input/output names which I then merge.

And with pseudo-sparse, is it reasonable to implement this soonish?

**Comment by AlexDBlack** [comment]
Unfortunately no - the graph implementation is not like that. It will initialize N different layers, with separate parameters. Even if you *could* force it to be one layer object after initialization, it wouldn't work (being stateful, and only the last gradients would be counted).

As for pseudo-sparse: I'd be hesitant to implement it just for this (I think this is the only use case it'd be good for), especially given we have ND4J sparse on the horizon...

**Comment by firasdib** [comment]
I gave the stack/unstack option a go, but unfortunately I don't think it works out for me.

The system takes several features, and every feature of the same type is fixed length. That is, my word features are of length `N` and char features are of length `M`. They are always this length. After they have been projected, they become `N'` and `M'` respectively, always.

However, looking at unstack, it would seem it assumes all are of the same size, which they are not in my case, and thus fails.

Did I miss anything here, or is there a workaround?
